src/JsonSchema/Constraints/SchemaConstraint.php

```diff
@@ -20,23 +23,70 @@
  */
 class SchemaConstraint extends Constraint
 {
+    const DEFAULT_SCHEMA_SPEC = 'http://json-schema.org/draft-04/schema#';
+
     /**
      * {@inheritdoc}
      */
     public function check(&$element, $schema = null, JsonPointer $path = null, $i = null)
     {
         if ($schema !== null) {
             // passed schema
-            $this->checkUndefined($element, $schema, $path, $i);
+            $validationSchema = $schema;
         } elseif ($this->getTypeCheck()->propertyExists($element, $this->inlineSchemaProperty)) {
             $inlineSchema = $this->getTypeCheck()->propertyGet($element, $this->inlineSchemaProperty);
             if (is_array($inlineSchema)) {
                 $inlineSchema = json_decode(json_encode($inlineSchema));
             }
             // inline schema
-            $this->checkUndefined($element, $inlineSchema, $path, $i);
+            $validationSchema = $inlineSchema;
         } else {
             throw new InvalidArgumentException('no schema found to verify against');
         }
+
+        // validate schema against whatever is defined in $validationSchema->$schema. If no
+        // schema is defined, assume self::DEFAULT_SCHEMA_SPEC (currently draft-04).
+        if ($this->factory->getConfig(self::CHECK_MODE_VALIDATE_SCHEMA)) {
+            if (!$this->getTypeCheck()->isObject($validationSchema)) {
+                throw new RuntimeException('Cannot validate non-object schema');
+            }
+            if ($this->getTypeCheck()->propertyExists($validationSchema, '$schema')) {
+                $schemaSpec = $this->getTypeCheck()->propertyGet($validationSchema, '$schema');
+            } else {
+                $schemaSpec = self::DEFAULT_SCHEMA_SPEC;
+            }
+
+            // preload standard schema specs
+            $preload = array(
+                'http://json-schema.org/draft-03/schema' => 'json-schema-draft-03.json',
+                'http://json-schema.org/draft-03/schema#' => 'json-schema-draft-03.json',
+                'http://json-schema.org/draft-04/schema' => 'json-schema-draft-04.json',
+                'http://json-schema.org/draft-04/schema#' => 'json-schema-draft-04.json'
+            );
+            $schemaStorage = $this->factory->getSchemaStorage();
+            foreach ($preload as $schemaID => $schemaFile) {
```

Comment:
couldn't that be moved inside the SchemaStorage class ? it would benefit everyone validating schemas (and protect us if something like https://github.com/json-schema-org/json-schema-org.github.io/issues/48 happen again)
for the record I now have a local copy of the http://json-schema.org/draft-04/schema# and would like to remove that if it's handled at the library level